Add checksum support

danielhodder opened the pull request:
This adds dependency integrity checking to the plugin. It's off by default (for compatibility), and only works with the JSON format option. As far as I can tell with the existing way of parsing maven artifacts there's no place to put the checksum without breaking maven's ability to parse the POM.

This uses Maven's view of the resolved artifact and does a SHA-256 sum on it and then adds that to the artifact metadata. Since I didn't wan to tunnel the configuration option all the way down to the `toString` method of `LockedDependency` it will always have a checsum component on the end of it now. This is identified by `@<checksum>`, a format taken from the container systems which use `image@ref` for this kind of thing.

The tests gave me a lot of trouble here since the MRN plugin does not seem to transfer the JAR files exactly as they are. There's a noticeable difference between the JAR files checked into the project and what appears in `target/local-repository`. The contents are the same but directory entries are removed, and it seems that the header meta-data is also changed. To work around this the new integration tests only test the integrity check matches that in the local registry, rather than the source file.

Issue: #49

vandmo commented:
Thanks a lot! I will have a closer look a this during the weekend.
Some quick notes now though.

It is possible to add additional information in pom.xml in a different XML namespace.

I think "integrity": "sha256-7r...Bg==", might be better. It will never make sense to add multiple checksums, only the one that is secure enough. It would be good if older versions of the plugin fails if it encounters a checksum that it doesn't understand. Saying something like "Encountered unsupported checksum format, consider using a later version of this plugin".

The master branch is intended for a 1.0 version of the plugin where configuration will be more flexible and integrity checking enabled by default.
```
<dependencySets>
  <dependencySet>
    <ignoreVersion>true</ignoreVersion>
    <ignoreIntegrity>true</ignoreIntegrity>
    <includes>
      <include>com.mycompany.something</include>
      ...
    </includes>
    <excludes>...</excludes>
  </dependencySet>
  <dependencySet>
     ....
  </dependencySet>
</dependencySets
```

You could target the 0.x branch if you want to as well, but there it will have to be 100% opt-in.

danielhodder commented:
I would ideally like to use this functionality soon-ish so if the 1.0 release doesn't have a soon-ish timescale then targeting the 0.x branch probably makes sense.

I will work on the rest of these changes this week and try and get them through as soon as I can.

danielhodder commented:
I've added a `sha256` header to the integrity strings, and added the verification of that. I have also made it write out that as part of a POM but I have no idea how to read that back in, without parsing the POM using a different XML mechanism. The existing maven parser, correctly, ignores the integrity elements but that means that data isn't available to be read back in at this point. which means while it _records_ the checksum, it can't actually check it right now.

vandmo commented on README.md:
```diff
@@ -85,6 +85,14 @@ Which lock file format to use, defaults to _json_.
 * _json_, lock file in JSON format, default filename is dependency-lock.json
 * _pom_, lock file in POM XML format, default filename is .dependeny-lock/pom.xml
 
+### dependencyIntegrityChecking
```
Maybe just checkIntegrity or just "integrity" since it both adds the entries and verifies them.

danielhodder replied:
changed to `checkIntegrity`

vandmo commented on src/it/projects/json-create-lock-file-with-integrity-checking/postbuild.groovy:
```diff
@@ -0,0 +1,31 @@
+import com.google.common.io.BaseEncoding
```
I follow https://google.github.io/styleguide/javaguide.html which sorts imports alphabetically.
You can use https://github.com/vandmo/google-java-format or similar to format the files but I can do it after merging as well.

danielhodder replied:
Ran the formatter over everything and commited the results.

vandmo commented on src/it/projects/pom-create-lock-file-with-integrity-checking/expected-pom.xml:
```diff
@@ -0,0 +1,163 @@
+  xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd"
+  xmlns:integrity="urn:se.vandmo.dependencylock.integrity">
```
I haven't seen urn:... used before but it kind of makes more sense then pointing to a http location that doesn't exist :+1: 
Should probably be just `urn:se.vandmo.dependencylock` though.

danielhodder replied:
Changed to that.

vandmo commented on src/it/projects/pom-create-lock-file-with-integrity-checking/expected-pom.xml:
```diff
@@ -0,0 +1,163 @@
+      <version>4.1.65.Final</version>
+      <type>jar</type>
+      <scope>compile</scope>
+      <integrity:checksum>sha256:E7679BA4F3B2FBEC1CD356380B056254FC4E9C0093EF7B083FBF75A79A6FE333</integrity:checksum>
```
NPM seems to use sha512 in package-lock.json for some reason. I haven't dug any deeper though, maybe it is because of length extension attacks? I am all for using a fast algorithm that is "secure enough" but is SHA256 that algorithm?

danielhodder replied:
If NPM uses SHA-512 then I think the sanest answer is to just copy what they do. I'll update to SHA-512 across this pr.

vandmo commented on src/main/java/se/vandmo/dependencylock/maven/Artifact.java:
```diff
@@ -1,18 +1,45 @@
+
+  static {
+    try {
+      SHA256_DIGEST = MessageDigest.getInstance("SHA-256");
```
SHA384 ? SHA512?
Is SHA256 secure enough?

danielhodder replied:
Will change to 512

vandmo reviewed (changes requested):
Thanks a lot! I will look dig into this deeper during the weekend. Seems like the XML might need to be parsed with something else but MavenXpp3Reader. Will need to bump jackson as well and make a new release.

vandmo commented:
I realized this was for the master branch which is for 1.0 just before merging. I have pushed some changes and I am working on parsing the pom file to get the integrity field. After that I can create a 0.0.commit-id release so you can try it out if you want to. Integrity checking should be enabled by default for the 1.0 release so there will be incompatible changes.

vandmo commented:
POM is now parsed "manually". I still have some work to do, but I am working actively on it so I am hoping to have a 1.0 or at least a 0.0.commit which will be close to the final 1.0 release.
I also encountered the strange behavior where mrm changes the jar file and the hash no longer matches. Did you ever find a solution to that @danielhodder ?

danielhodder commented:
If you mean checksums of the jars that are exposed via MRN then yes. I ran into this in the JSON format as well https://github.com/vandmo/dependency-lock-maven-plugin/pull/62/files#diff-07cefa6399119458dd8677039ac63194cee36f4342ebd161550870ba2ff1fca7R16-R20. Somehow the process which copies the jars from `src/it/mrn` to `target/local-repo` seems to re-write the entities in the jar. I suspect something it iterating a ZipFileStream, and re-writeing it.

This is what the entries in the `src/it/mrn/repository/leaf-1.0.jar` looks like

```
unzip -l src/it/mrm/repository/leaf-1.0.jar
Archive:  src/it/mrm/repository/leaf-1.0.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  12-31-2008 22:16   META-INF/
       71  12-31-2008 22:16   META-INF/MANIFEST.MF
---------                     -------
       71                     2 files
```

While this is what it looks like in the `target/local-repository`

```
unzip -l target/local-repo/se/vandmo/testing/leaf/1.0/leaf-1.0.jar
Archive:  target/local-repo/se/vandmo/testing/leaf/1.0/leaf-1.0.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
       90  10-06-2022 09:15   META-INF/MANIFEST.MF
---------                     -------
       90                     1 file
```

vandmo commented:
Apparently MRM creates a new JAR for the manifest even when a jar is supplied. Anyhow, turns out MRM isn't needed. I created a setup-dependencies module which will run first and run install-file for the jar files.
